File: J9Ch/src/S_7_Follow_Up/Required_14/_139_Subarray_Sum_Closest_Medium.cpp
/** 139 Subarray Sum Closest
 * Medium

 * Created by tianhuizhu on 6/28/17.
 */


class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return: A list of integers includes the index of the first number
     *          and the index of the last number
     */
    struct node {
        node(int _value, int _pos):value(_value), pos(_pos) {}
        int value, pos;
        bool operator<(const node &o) const{
            return (value < o.value || value == o.value && pos < o.pos);
        }
    };
    vector<int> subarraySumClosest(vector<int> nums){
        // write your code here
        vector<node> s;
        vector<int> results(2);
        s.push_back(node(0,-1));
        int sum = 0, len = nums.size();
        for (int i = 0; i < len ; ++i) {
            sum += nums[i];
            s.push_back(node(sum, i));
        }
        sort(s.begin(), s.end());
        len = s.size();
        int ans = 0x7fffffff;
        for (int i = 0; i < len-1; ++i)
            if (abs(s[i+1].value - s[i].value) < ans) {
                ans = abs(s[i+1].value - s[i].value);
                results[0] = min(s[i].pos, s[i+1].pos)+1;
                results[1] = max(s[i].pos, s[i+1].pos);
            }
        return results;
    }
};

File: J9Ch/src/S_7_Follow_Up/Required_14/python_cpp/_139_Subarray_Sum_Closest_Medium.py
'''/** 139 Subarray Sum Closest
 * Medium

 * Created by tianhuizhu on 6/28/17.
 */'''


class Node:
    def __init__(self, _value, _pos):
        self.value = _value
        self.pos = _pos
    def __cmp__(self, other):
        if self.value == other.value:
            return self.pos - other.pos
        return self.value - other.value
class Solution:
    """
    @param nums: A list of integers
    @return: A list of integers includes the index of the first number
             and the index of the last number
    """


    def subarraySumClosest(self, nums):
        # write your code here
        s = []
        s.append(Node(0, -1))
        sum = 0
        for x in xrange(len(nums)):
            sum += nums[x]
            s.append(Node(sum, x))

        s = sorted(s)
        results= [0,0]
        ans = 1000000000000
        for i in xrange(len(s)-1):
            if s[i+1].value - s[i].value < ans or \
                                            s[i+1].value - s[i].value == ans and \
                                            min(s[i+1].pos, s[i].pos) + 1 < results[0]:
                ans = s[i+1].value - s[i].value
                results[0] = min(s[i+1].pos, s[i].pos) + 1
                results[1] = max(s[i+1].pos, s[i].pos)

        return results